### ISSUE TITLE
auth_roles_test.TestAuthRoles.test_list_permissions AssertionError: the current permissions do not meet expectations

### ISSUE THREAD

**Issue by bhalevy** [scylladb/scylladb#13973]
The test fails from time to time, for example https://jenkins.scylladb.com/view/master/job/scylla-master/job/dtest-daily-release/251/testReport/auth_roles_test/TestAuthRoles/Run_Dtest_Parallel_Cloud_Machines___FullDtest___full_split000___test_list_permissions/
```
self = <auth_roles_test.TestAuthRoles object at 0x7fa760622350>

    def test_list_permissions(self):
        self.prepare()
        cassandra = self.get_session(user='cassandra', password='cassandra')
        cassandra.execute("CREATE KEYSPACE ks WITH replication = {'class':'SimpleStrategy', 'replication_factor':1}")
        cassandra.execute("CREATE TABLE ks.cf (id int primary key, val int)")
        cassandra.execute("CREATE ROLE mike WITH PASSWORD = '12345' AND SUPERUSER = false AND LOGIN = true")
        cassandra.execute("CREATE ROLE role1")
        cassandra.execute("CREATE ROLE role2")
        cassandra.execute("GRANT SELECT ON table ks.cf TO role1")
        cassandra.execute("GRANT ALTER ON table ks.cf TO role2")
        cassandra.execute("GRANT MODIFY ON table ks.cf TO mike")
        cassandra.execute("GRANT ALTER ON ROLE role1 TO role2")
        cassandra.execute("GRANT role1 TO role2")
        cassandra.execute("GRANT role2 TO mike")
    
        expected_permissions = [("mike", "<table ks.cf>", "MODIFY"),
                                ("role1", "<table ks.cf>", "SELECT"),
                                ("role2", "<table ks.cf>", "ALTER"),
                                ("role2", "<role role1>", "ALTER")]
        expected_permissions.extend(data_resource_creator_permissions('cassandra', '<keyspace ks>', support_func=False))
        expected_permissions.extend(data_resource_creator_permissions('cassandra', '<table ks.cf>'))
        expected_permissions.extend(role_creator_permissions('cassandra', '<role mike>'))
        expected_permissions.extend(role_creator_permissions('cassandra', '<role role1>'))
        expected_permissions.extend(role_creator_permissions('cassandra', '<role role2>'))
    
>       self.assert_permissions_listed(expected_permissions, cassandra, "LIST ALL PERMISSIONS")

auth_roles_test.py:515: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

expected = [('mike', '<table ks.cf>', 'MODIFY'), ('role1', '<table ks.cf>', 'SELECT'), ('role2', '<table ks.cf>', 'ALTER'), ('role2', '<role role1>', 'ALTER'), ('cassandra', '<keyspace ks>', 'SELECT'), ('cassandra', '<keyspace ks>', 'MODIFY'), ...]
session = <cassandra.cluster.Session object at 0x7fa76037bd50>
query = 'LIST ALL PERMISSIONS'

    @staticmethod
    def assert_permissions_listed(expected, session, query):
        rows = session.execute(query)
        perms = [(str(r.role), str(r.resource), str(r.permission)) for r in rows]
>       assert sorted(expected) == sorted(perms), "the current permissions do not meet expectations"
E       AssertionError: the current permissions do not meet expectations
E       assert [('cassandra', '<keyspace ks>', 'ALTER'),\n ('cassandra', '<keyspace ks>', 'AUTHORIZE'),\n ('cassandra', '<keyspace ks>', 'CREATE'),\n ('cassandra', '<keyspace ks>', 'DROP'),\n ('cassandra', '<keyspace ks>', 'MODIFY'),\n ('cassandra', '<keyspace ks>', 'SELECT'),\n ('cassandra', '<role mike>', 'ALTER'),\n ('cassandra', '<role mike>', 'AUTHORIZE'),\n ('cassandra', '<role mike>', 'DROP'),\n ('cassandra', '<role role1>', 'ALTER'),\n ('cassandra', '<role role1>', 'AUTHORIZE'),\n ('cassandra', '<role role1>', 'DROP'),\n ('cassandra', '<role role2>', 'ALTER'),\n ('cassandra', '<role role2>', 'AUTHORIZE'),\n ('cassandra', '<role role2>', 'DROP'),\n ('cassandra', '<table ks.cf>', 'ALTER'),\n ('cassandra', '<table ks.cf>', 'AUTHORIZE'),\n ('cassandra', '<table ks.cf>', 'DROP'),\n ('cassandra', '<table ks.cf>', 'MODIFY'),\n ('cassandra', '<table ks.cf>', 'SELECT'),\n ('mike', '<table ks.cf>', 'MODIFY'),\n ('role1', '<table ks.cf>', 'SELECT'),\n ('role2', '<role role1>', 'ALTER'),\n ('role2', '<table ks.cf>', 'ALTER')] == [('cassandra', '<all functions in ks>', 'ALTER'),\n ('cassandra', '<all functions in ks>', 'AUTHORIZE'),\n ('cassandra', '<all functions in ks>', 'CREATE'),\n ('cassandra', '<all functions in ks>', 'DROP'),\n ('cassandra', '<all functions in ks>', 'EXECUTE'),\n ('cassandra', '<keyspace ks>', 'ALTER'),\n ('cassandra', '<keyspace ks>', 'AUTHORIZE'),\n ('cassandra', '<keyspace ks>', 'CREATE'),\n ('cassandra', '<keyspace ks>', 'DROP'),\n ('cassandra', '<keyspace ks>', 'MODIFY'),\n ('cassandra', '<keyspace ks>', 'SELECT'),\n ('cassandra', '<role mike>', 'ALTER'),\n ('cassandra', '<role mike>', 'AUTHORIZE'),\n ('cassandra', '<role mike>', 'DROP'),\n ('cassandra', '<role role1>', 'ALTER'),\n ('cassandra', '<role role1>', 'AUTHORIZE'),\n ('cassandra', '<role role1>', 'DROP'),\n ('cassandra', '<role role2>', 'ALTER'),\n ('cassandra', '<role role2>', 'AUTHORIZE'),\n ('cassandra', '<role role2>', 'DROP'),\n ('cassandra', '<table ks.cf>', 'ALTER'),\n ('cassandra', '<table ks.cf>', 'AUTHORIZE'),\n ('cassandra', '<table ks.cf>', 'DROP'),\n ('cassandra', '<table ks.cf>', 'MODIFY'),\n ('cassandra', '<table ks.cf>', 'SELECT'),\n ('mike', '<table ks.cf>', 'MODIFY'),\n ('role1', '<table ks.cf>', 'SELECT'),\n ('role2', '<role role1>', 'ALTER'),\n ('role2', '<table ks.cf>', 'ALTER')]
E         At index 0 diff: ('cassandra', '<keyspace ks>', 'ALTER') != ('cassandra', '<all functions in ks>', 'ALTER')
E         Right contains 5 more items, first extra item: ('cassandra', '<table ks.cf>', 'SELECT')
E         Full diff:
E           [
E         -  ('cassandra',
E         -   '<all functions in ks>',
E         -   'ALTER'),
E         -  ('cassandra',
E         -   '<all functions in ks>',
E         -   'AUTHORIZE'),
E         -  ('cassandra',
E         -   '<all functions in ks>',
E         -   'CREATE'),
E         -  ('cassandra',
E         -   '<all functions in ks>',
E         -   'DROP'),
E         -  ('cassandra',
E         -   '<all functions in ks>',
E         -   'EXECUTE'),
E            ('cassandra',
E             '<keyspace ks>',
E             'ALTER'),
E            ('cassandra',
E             '<keyspace ks>',
E             'AUTHORIZE'),
E            ('cassandra',
E             '<keyspace ks>',
E             'CREATE'),
E            ('cassandra',
E             '<keyspace ks>',
E             'DROP'),
E            ('cassandra',
E             '<keyspace ks>',
E             'MODIFY'),
E            ('cassandra',
E             '<keyspace ks>',
E             'SELECT'),
E            ('cassandra',
E             '<role mike>',
E             'ALTER'),
E            ('cassandra',
E             '<role mike>',
E             'AUTHORIZE'),
E            ('cassandra',
E             '<role mike>',
E             'DROP'),
E            ('cassandra',
E             '<role role1>',
E             'ALTER'),
E            ('cassandra',
E             '<role role1>',
E             'AUTHORIZE'),
E            ('cassandra',
E             '<role role1>',
E             'DROP'),
E            ('cassandra',
E             '<role role2>',
E             'ALTER'),
E            ('cassandra',
E             '<role role2>',
E             'AUTHORIZE'),
E            ('cassandra',
E             '<role role2>',
E             'DROP'),
E            ('cassandra',
E             '<table ks.cf>',
E             'ALTER'),
E            ('cassandra',
E             '<table ks.cf>',
E             'AUTHORIZE'),
E            ('cassandra',
E             '<table ks.cf>',
E             'DROP'),
E            ('cassandra',
E             '<table ks.cf>',
E             'MODIFY'),
E            ('cassandra',
E             '<table ks.cf>',
E             'SELECT'),
E            ('mike',
E             '<table ks.cf>',
E             'MODIFY'),
E            ('role1',
E             '<table ks.cf>',
E             'SELECT'),
E            ('role2',
E             '<role role1>',
E             'ALTER'),
E            ('role2',
E             '<table ks.cf>',
E             'ALTER'),
E           ]

auth_roles_test.py:1313: AssertionError
```

**Comment by mykaul** [comment]
If we think this is caused by the recent permissions changes, I suggest reverting the changes first, fixing (inc. the tests), then re-adding them.

**Comment by wmitros** [comment]
The dtest that is failing in this case is was not updated when the UDF permissions were implemented in Scylla. This should be fixed with https://github.com/scylladb/scylla-dtest/pull/3179

**Comment by bhalevy** [comment]
> The dtest that is failing in this case is was not updated when the UDF permissions were implemented in Scylla. This should be fixed with [scylladb/scylla-dtest#3179](https://github.com/scylladb/scylla-dtest/pull/3179)

Unfortunately it still fails at scylladb/scylla-dtest@194a0225e02d785bcdf8e58246efa52a6f928c3b that includes https://github.com/scylladb/scylla-dtest/commit/ab7ebdb4585058e3474c0cdf50f6a54f5a9a8a56.
See https://jenkins.scylladb.com/view/master/job/scylla-master/job/dtest-daily-release/270/testReport/auth_roles_test/TestAuthRoles/Run_Dtest_Parallel_Cloud_Machines___FullDtest___full_split000___test_list_permissions/

**Comment by bhalevy** [comment]
ping, the test is still failing on a regular basis.

**Comment by bhalevy** [comment]
Cc @DoronArazii this test failed last 30 dtest-daily-release runs

**Comment by mykaul** [comment]
@wmitros - please take a look

**Comment by DoronArazii** [comment]
If it will stay as P2, it will take time until Wojciech will get to it.
I'm raising the priority.

**Comment by wmitros** [comment]
Sorry for the late reply, I just came back from vacations
The test first started failing after https://github.com/scylladb/scylladb/pull/13814 but got fixed in https://github.com/scylladb/scylla-dtest/pull/3179. However, https://github.com/scylladb/scylladb/pull/13814 got reverted and without it https://github.com/scylladb/scylla-dtest/pull/3179 is causing the current errors. I think we should just revert https://github.com/scylladb/scylla-dtest/pull/3179 for now.
There's actually nothing wrong with the reverted patch. It just uncovered another issue (https://github.com/scylladb/scylla-enterprise/issues/3058) that is currently getting fixed. When we finish that fix we will have a couple of re-reverts to do, but it's worth unblocking the dtest-daily job.
@fruch can you help with reverting https://github.com/scylladb/scylla-dtest/pull/3179? Or should I prepare another patch?

**Comment by fruch** [comment]
@wmitros 

other option is to add a require marker to test with `https://github.com/scylladb/scylla-enterprise/issues/3058`

reverting would be with a PR anyhow at this stage.

I'm not quite understand how the test would stop failing if we revert this change... wouldn't still do a check that would be failing ?

**Comment by wmitros** [comment]
> @wmitros
> 
> other option is to add a require marker to test with `https://github.com/scylladb/scylla-enterprise/issues/3058`

A marker also sounds good, but probably with https://github.com/scylladb/scylladb/issues/13747 because https://github.com/scylladb/scylla-enterprise/issues/3058 won't fix the issue on its own, it's just a prerequisite for https://github.com/scylladb/scylladb/issues/13747. Still, the test case isn't really correlated to either of these issues so I'm not sure if it's the best approach

> reverting would be with a PR anyhow at this stage.
> 
> I'm not quite understand how the test would stop failing if we revert this change... wouldn't still do a check that would be failing ?

A check would still be performed, but the comparison performed in it would start succeeding. Currently the list of expected permissions that we're checking includes permissions on functions, and reverting https://github.com/scylladb/scylla-dtest/pull/3179 would remove them. (https://github.com/scylladb/scylladb/pull/13814 made it so that the actual permissions returned by Scylla started including permissions on functions. After reverting it, Scylla no longer returns permissions on functions)

**Comment by bhalevy** [comment]
Can we just move the test from scylla-dtest to scylla/test/cql-test and then it would be synchronized with the scylla branch it lives in?

Plus, we'll get some cookie points from @nyh ;-)

**Comment by wmitros** [comment]
> Can we just move the test from scylla-dtest to scylla/test/cql-test and then it would be synchronized with the scylla branch it lives in?
> 
> Plus, we'll get some cookie points from @nyh ;-)

I'd also prefer the test to be in `cql-pytest`, however that would require a bit more work because the test uses more auxiliary functions that are already in dtests.
For now I posted the revert at https://github.com/scylladb/scylla-dtest/pull/3281 to stop the job from failing
